### PR TITLE
Add nether biomes API

### PR DESCRIPTION
**Pull request by jaskarth**
Add a new class and mixin to allow modders to inject their own biomes into the nether in a compatible manner.

### CODE REVIEW

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:17]: licence

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: licencE

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:17]: lIcenCE

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java:32]: good

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: nice

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (valoeghese) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:37]: approve

**Review by valoeghese — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (Earthcomputer) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java]: Wouldn't this be better as a `@ModifyArg`?

**Review by Earthcomputer — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (jaskarth) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java]: Done

**Review by valoeghese — APPROVED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]:
```suggestion
 * API that exposes the internals of Minecraft's nether biome code.
```

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]:
```suggestion
	 * @param biome The biome to add. Must not be null.
```

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: Add an `Objects.requireNonNull` check here

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java]: comments on the preceding line are more readable imo
```suggestion
		// the set in the parameter is immutable, so we construct our own
		Set<Biome> newSet = new HashSet<>(set);
```

File: fabric-biomes-v1/src/main/resources/fabric-biomes-v1.mixins.json
Comment (coderbot16) [inline fabric-biomes-v1/src/main/resources/fabric-biomes-v1.mixins.json]: Maintain alphabetical order please

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: Consider wrapping this with `Collections.unmodifiableSet`.

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Some easy fixes.

**Review by Prospector — APPROVED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java
Comment (liach) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheNetherDimension.java]: Unused import

**Review by liach — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: Perhaps elaborate on which settings and have a `{@link Biome.Settings}` as well

**Review by coderbot16 — CHANGES_REQUESTED** [review]
One final review, I promise! I think we should just clarify to API users what settings they need to tweak.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (liach) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: Need to detail that you need to set up the biome noise in settings properly or the biomes have literally zero chance to generate in the nether

Comment (jaskarth) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: I think i will add some basic information and then write some detailed explanation on the wiki

**Review by liach — APPROVED** [review]
Build passes. Style is good.

**Review by coderbot16 — APPROVED** [review]

**Comment by coderbot16** [comment]
@Earthcomputer can we get a second review from you? Just want to clean up the reviews list :p

**Review by Earthcomputer — APPROVED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (Juuxel) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]:
```suggestion
	 * @see Biome.MixedNoisePoint
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (coderbot16) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: If we're applying Juuz's comment,
```suggestion
	 * Adds a biome to the Nether generator. Biomes must set their own noise values in the {@link Biome.MixedNoisePoint} class for the biome to properly generate.
```

**Comment by coderbot16** [comment]
I think this is ready for a final review from Player/modmuss. Can @Juuxel give this another look?

**Comment by coderbot16** [comment]
Almost forgot, we should probably increment the version number to 0.2.0. I'm not sure how this will interact with #491, or if we'll need to bump it to 0.3.0 in the other PR.

**Comment by Prospector** [comment]
Bump to 0.2.0 in both and unless they're merged at the same time, bump the one that was merged last to 0.3.0

**Review by Juuxel — APPROVED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (dblsaiko) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: Everywhere else in API (afaik) this class would be an interface with an INSTANCE field that NetherBiomesImpl implements (see e.g. net.fabricmc.fabric.api.client.model.ModelLoadingRegistry.) Why is this different here?

Not that it makes any real difference, but it would be more consistent.

**Review by dblsaiko — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (jaskarth) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: I wanted to be consistent with the rest of the biome API which follows the same pattern that I used.

Comment (SuperCoder7979) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: I wanted to be consistent with the rest of the biome API which follows the same pattern that I used.

Comment (Prospector) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: Hm, yeah I didn't even notice we did the OverworldBiomes one any differently. Also I noticed in OverworldBiomes we had the class as `final`, maybe put that here too?

Comment (jaskarth) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: fixed 👍

Comment (dblsaiko) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java]: Yeah, if this style exists somewhere previously in API, then fair enough.

**Review by dblsaiko — APPROVED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (modmuss50) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:26]: Why not an interface like most of the other apis?

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (modmuss50) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: Does this contain vanilla biomes?

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (modmuss50) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: I guess its not an issue, as its not exposed via the api, ingore me.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java
Comment (liach) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/NetherBiomes.java:26]: Should be fine as you can just change the method body as long as the method names and descriptors stay intact

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (liach) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: No, just extra biomes to be added

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (modmuss50) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: Why make this a class and not add it to `InternalBiomeData`?

**Comment by modmuss50** [comment]
Module also needs to be marked to only run on 1.16+ versions

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java
Comment (jaskarth) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/NetherBiomesImpl.java]: That is a great question if i'm being honest

**Review by modmuss50 — APPROVED** [review]
Implement #499 and then I think this is good to merge.

**Review by YanisBft — APPROVED** [review]

**Pull request merged**
